confmap/provider/googlesecretsprovider/provider_test.go

```diff
@@ -0,0 +1,127 @@
+	if m.err != nil {
+		return nil, m.err
+	}
+	return &secretmanagerpb.AccessSecretVersionResponse{
+		Payload: &secretmanagerpb.SecretPayload{
+			Data: []byte(m.secretString),
+		},
+	}, nil
+}
+
+func (m *mockSecretsManagerClient) Close() error {
+	return nil
+}
+
+func TestProvider_Retrieve_Success(t *testing.T) {
+	tests := []struct {
+		name              string
+		uri               string
+		testSecretManager *mockSecretsManagerClient
+		wantSecret        string
+	}{
+		{
+			name: "Happy path: valid uri, secret entry exists and is accessible",
+			uri:  schemeName + ":projects/my-project/secrets/test-secret-id/versions/1",
+			testSecretManager: &mockSecretsManagerClient{
+				err:          nil,
+				secretString: "test-secret-value",
+			},
+			wantSecret: "test-secret-value",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			testProvider := &provider{
+				client: tc.testSecretManager,
+			}
+			defer testProvider.client.Close()
+			gotSecret, err := testProvider.Retrieve(context.Background(), tc.uri, nil)
+			if err != nil {
+				t.Errorf("%v: Retrieve() gotError = %v, want nil error", tc.name, err)
+			}
+			gotSecretString, err := gotSecret.AsString()
+			if err != nil {
+				t.Errorf("%v: failed to retrieve the string value of the secret, error: %v", tc.name, err)
+			}
+			if gotSecretString != tc.wantSecret {
+				t.Errorf("%v: Retrieve() gotSecret = %v, want %v", tc.name, gotSecret, tc.wantSecret)
+			}
+		})
+	}
+}
+
+func TestProvider_Retrieve_Failure(t *testing.T) {
+	tests := []struct {
+		name              string
+		uri               string
+		testSecretManager *mockSecretsManagerClient
+	}{
+		{
+			name: "Invalid scheme",
+			uri:  "invalidscheme" + ":projects/my-project/secrets/test-secret-id/versions/1",
+			testSecretManager: &mockSecretsManagerClient{
+				err:          errors.New("invalid scheme"),
+				secretString: "test-secret-value",
+			},
+		},
+		{
+			name: "secret entry does not exist in the secret manager",
+			uri:  schemeName + ":projects/my-project/secrets/non-existent/versions/1",
+			testSecretManager: &mockSecretsManagerClient{
+				err:          errors.New("secret entry does not exist"),
+				secretString: "test-secret-value",
+			},
+		},
+		{
+			name: "invalid secret name",
+			uri:  schemeName + ":projects/my-project/invalid-secret/test-secret-id/versions/1",
+			testSecretManager: &mockSecretsManagerClient{
+				err:          errors.New("secret name is invalid"),
+				secretString: "test-secret-value",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			testProvider := &provider{
+				client: tc.testSecretManager,
+			}
+			defer testProvider.client.Close()
+			_, err := testProvider.Retrieve(context.Background(), tc.uri, nil)
```

Comment:
This isn't actually testing the types of errors; if all test cases happened to be failing with the same error due to some bug, these wouldn't be caught. We should change these tests to actually use [`require.ErrorIs`](https://pkg.go.dev/github.com/stretchr/testify/require#ErrorIs) and assert the error type is the expected one for this test case.

Reply:
The mock secret manager implementation has been updated to return a plaintext secret only when the provided `secretName` exists in a predefined map. Otherwise, a "not found" error is returned.

The unit test only verifies whether errors are expected or not; it does not verify if the correct error code is returned. That's up to the caller of Retrieve() to verify and act upon. Retrieve() only forwards the errors from the secret manager to the caller.